### PR TITLE
Right2Drive/AT-57/nav-tabs

### PR DESCRIPTION
**Pull request by Right2Drive**
Implemented TabManager component & redux state for tabs

### CODE REVIEW

File: src/store/modules/tabs/types.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/** Tab module state */
+export interface TabState {
+  tabs: StoreTabList;
+  activeTab: Tab | null;
+}
+
+/*----------------------- Actions -----------------------*/
+
+/** Tab module action */
+export type Action = {
```
Comment (beaesguerra) [inline src/store/modules/tabs/types.ts]: Nvm my question about action types. Maybe name this TabAction so it is clear?

File: src/store/types.ts [hunk 2]
```diff
@@ -1,13 +1,17 @@
 import { AppState, Action as AppAction } from './modules/app/types';
 import { NavState, Action as NavAction } from './modules/nav/types';
+import { TabState, Action as TabsAction } from './modules/tabs/types';
```
Comment (beaesguerra) [inline src/store/types.ts]: If the actions were not all named "Action" then we could avoid aliasing them each time we import different kinds of actions

File: src/view/components/TabManager/index.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+/** TabManager OwnProps */
+interface OwnProps {}
+
+interface StoreProps {
+  activeTab: string;
+}
+
+interface DispatchProps extends DispatchProp<State> {
+  setTabs: (tabs: ElementTabList, activeTab: ElementTab) => void;
+  setActiveTab: (activeTab: ElementTab) => void;
```
Comment (beaesguerra) [inline src/view/components/TabManager/index.tsx]: Do you think it's the tab managers responsibility to clear tabs? Maybe the lifecycle after it is unmounted could clear the tabs so that they disappear when there is a change in page

File: src/view/components/TabManager/index.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+const mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps> = (dispatch) => {
+  return {
+    setActiveTab: tab => dispatch(tabActions.setActiveTab(tab)),
+    setTabs: (tabs, activeTab) => dispatch(tabActions.setTabs(toTabs(tabs), activeTab)),
+  };
+};
+
+/** Creation utility for a TabManager */
+export default function createTabManager(tabs: ElementTabList) {
+  return connect(
```
Comment (beaesguerra) [inline src/view/components/TabManager/index.tsx]: So how do we use this tabmanager? Do we extend it still?

File: src/store/modules/app/actions.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import { Action } from './types';
 
-const actions = {
+const appActions = {
```
[the remark this answers is not in the document]
Comment (Right2Drive) [reply src/store/modules/app/actions.ts]: It's the entire module that must be renamed, but yes I agree. Enoch brought this up regarding agriboard and I have changed it there, and it is absolutely my mistake as I steered Muaz to use this state schema for redux. I think this should be it's own issue though

File: src/store/types.ts (hunk 2 above)
Comment (Right2Drive) [reply src/store/types.ts]: I will address all of your concerns with the names of Props and Actions in this comment (I can't respond to your general comment for the PR).

I agree that being verbose is good, and here is why I believe that my decision to use the names I did is actually more readable than if I were to use `TabsAction` or `TabManagerOwnProps`. 

1) Typically these interfaces are being used in the context of the component or module that it belongs to. For example, in all likelihood the only time `TabManagerProps` will be referenced is inside of this component. So when you instead see `Props` you instinctively know that these are the props for the component whose file you have opened. `TabManagerProps` is, in my opinion, adding unnecessary verbosity in this situation that imo actually hinders readability, not improves it.
2)  Our linting rules specify a set length of line that is permitted before we should be splitting a single line statement across multiple lines. These long names such as `TabManagerOwnProps` make it more likely to have to split a line such as the following into multiple lines:

```
const mapStateToProps: MapStateToProps<StoreProps, OwnProps, State> = (state, ownProps) => {

// becomes

const mapStateToProps: MapStateToProps<TabManagerStoreProps, TabManagerOwnProps, State> = (state, ownProps) => {
```

And being that they are types, it's function statements like this that are more likely to suffer from having many different types and therefore being split across multiple lines. Personally I find this far more difficult to read:
```
function myFunc(
  arg1,
  arg2,
  arg3,
) {

}
```

I would far prefer that we use these shorter names and use _name aliases_ as was done in this file, and enforce the usage of these aliases when importing props, actions etc from another file outside the current module or component. I know this is largely a matter of opinion, but maybe some of these points ring true to you as well?

File: src/view/components/TabManager/index.tsx (hunk 3 above)
Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: Yeah it is I was thinking about this on my way home and realized I had never called clear. Good catch :+1:

File: src/view/components/TabManager/index.tsx (hunk 4 above)
Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: # TLDR;
So long story short no. This utility will actually create the class for you. You just pass in the tabs that you want your component to offer, and the function will return a class that you export. This would look similar to the following:
```
const spinnerTabs: ElementTabList = [
  {
    element: () => <SpinnerPage />,
    name: 'Test',
  },
  {
    element: () => <SpinnerPage />,
    name: 'Hello',
  },
];

const SpinnerTabManager = createTabManager(spinnerTabs);

export default MyNewTabManager;
```

# Long Story

## Problem 1
I cannot pass in an abstract class into the `react-redux` `connect` function. This has the implication that anyone not understanding how the base class `TabManager` worked would be tempted to instantiate it directly, which Typescript would not catch but would cause runtime errors, as originally the constructor had the following as a definition:
`(props: Props, tabs: ElementTabList) => void`
And there is no way (as far as I know) to pass in arguments to the constructor using `JSX`. 

## Problem 2
It would seem that after extending the redux-connected class that `connect` returns, there was no need to call `super(props, tabs)` in the constructor of the child class. Not doing this would again result in a bunch of runtime errors, so I'm kind of shocked that Typescript does not enforce it. I suspect if it's an issue with types, it is probably from issues within the `react-redux` typings and not Typescript itself. I'm not even sure if extending the class that `connect` returns actually works either tbh.

## Solution
Seeing as a class is just a function with some syntactic sugar, it occurred to me I could simply create a factory function that creates the class for you with simply one argument, the tabs. I know it looks bizarre in comparison to any other language, but I think it's still very true to the nature of Javascript and am happy I stumbled upon this.

PS: sorry for the essay

File: src/view/components/TabManager/index.tsx (hunk 3 above)
Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: I will actually make this change in my other PR if that's okay? I didn't realize I was going to have to touch TabManager again in order to actually render the tabs... but I did have to

File: src/store/types.ts (hunk 2 above)
Comment (Right2Drive) [reply src/store/types.ts]: PS: I can't take credit for the naming conventions `OwnProps` etc as I stole it from `react-redux` itself, but thank you for the compliment :)

If we do want to make it more verbose that they are types, I think a good solution to this would be enforcing an `I` prefix on interfaces and a `T` prefix on types. I have just been avoiding doing that because in vscode I don't typically need it as it will tell me what it is.

File: src/view/components/TabManager/index.tsx (hunk 4 above)
Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: I also need to document this function far better though

Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: In my new PR it now looks like this:

```
/**
 * Create a React Component class to manage the navigation between provided tabs
 *
 * The first tab in the 'tabs' list will be rendered on the far left, and the last
 * will be rendered on the far right.
 *
 * @example
 * const ExampleTabManager = createTabManager(exampleTabs);
 * export default ExampleTabManager
 *
 * @param {ElementTabList} [tabs] The tabs this component should render
 * @returns {React.ComponentClass}
 */
export default function createTabManager(tabs: ElementTabList) {
  return connect(
    mapStateToProps,
    mapDispatchToProps,
  )(createSimpleComponent(tabs));
}
```

File: src/view/components/TabManager/index.tsx (hunk 3 above)
Comment (Right2Drive) [reply src/view/components/TabManager/index.tsx]: Fixed in my other PR

File: src/store/modules/tabs/types.ts (hunk 1 above)
Comment (Right2Drive) [reply src/store/modules/tabs/types.ts]: Addressed in other reply

File: src/store/modules/tabs/reducer.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { Reducer } from 'redux';
+
+import { TabState, Action } from './types';
+import initialState from './state';
+
+const reducer: Reducer<TabState> = (state = initialState, action: Action) => {
```
[the remark this answers is not in the document]
Comment (Right2Drive) [reply src/store/modules/tabs/reducer.ts]: Responded in other comments, although the reducer name change might be a good idea considering it is only imported in one other location (the main reducer).

File: src/store/modules/app/actions.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (beaesguerra) [reply src/store/modules/app/actions.ts]: okay! Please make an issue to track this :)

File: src/store/types.ts (hunk 2 above)
Comment (beaesguerra) [reply src/store/types.ts]: It makes sense that you are mostly using it in the same context, but there are still cases where it is used elsewhere (like this file for example). I guess my concern would be that since we are aliasing it, someone might alias it another way in another file, and this could lead to some confusion because different files will call it different things. I know this is really trivial and it may be just me, but I don't usually read the imports when I'm looking through code so it would take me longer to realize that they were the same thing

Comment (Right2Drive) [reply src/store/types.ts]: How about a solution that is the best of both worlds?

```
/** TabManager OwnProps */
interface OwnProps {}

interface StoreProps {
  activeTab: string;
}

interface DispatchProps extends DispatchProp<State> {
  setTabs(tabs: ElementTabList, activeTab: ElementTab): void;
  setActiveTab(activeTab: ElementTab): void;
  clearTabs(): void;
}

interface OwnState {}

/** TabManager props */
type Props = OwnProps & StoreProps & DispatchProps;

// Here <-----
export interface TabManagerOwnState extends OwnState {}
export type TabManagerProps = Props;
```

File: src/store/modules/app/actions.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (beaesguerra) [reply src/store/modules/app/actions.ts]: link it here too so we know where it came from :)

File: src/store/types.ts (hunk 2 above)
Comment (beaesguerra) [reply src/store/types.ts]: sure that's fine. not sure if it's too redundant but maybe we can ask the rest of the team later!  
what's the difference between interface and type?
